### PR TITLE
Arbitrary data publish

### PR DESCRIPTION
**Pull request by stuthedew**
I added a method to publish arbitrary data in addition to the publish methods currently provided (int32_t, double, uint32_t and const char). I've also included a new example sketch showing how to use it. I reworked the backend of the library a little to unify things a little and eliminate redundant code, but I've tested it using Adafruit IO with all existing publish methods, and nothing seems broken! It should (fingers crossed) be a drop-in replacement for the current library.

img_2667

My motivation for this was wanting to be able to include metadata to a post.
Specifically, I was playing around with a [Teviso RD3024 radiation sensor](http://www.teviso.com/en/products/radiation-sensor-rd3024.htm) hooked up to your Huzzah ESP8266, and a salvaged Americium radiation source from a smoke detector, at varying distances from the sensor. I wanted a way to associate the collection time, and distance between the source and sensor with the actual radiation reading itself. I wrote a python-based daemon running on a Raspberry Pi (similar to the example mentioned above) to subscribe to the feed (Mosquito-based, also on Raspberry Pi), parse the data and insert it into a MySQL database running on the Pi as well. I then used Tableau on my Macbook to connect to the Pi/MySQL database and graph the data.

rad vs distance

Adafruit's libraries and resources are absolutely awesome, and I'm so appreciative of all the work you guys do. Being an entirely self-taught with no formal coding background (I'm a medical student), I regularly look at the work you guys do to get an idea of what quality coding looks like for my own projects. While I can't promise that this reflects the level of coding you regularly produce, I just thought I would pass this along it might be useful to someone. Thanks!

### CODE REVIEW

**Comment by ladyada** [comment]
hiya thanks for the submission - there's a *ton* of mini commits, which makes it very difficult to analyze what was added or changed, somtimes it seems you just changed a variable name or added/removed whitespace. if you could create a single commit with only the relevant changes it would help me tremendously in testing and adding this in, thx! :)

**Comment by ladyada** [comment]
@toddtreece  kindly showed me https://github.com/adafruit/Adafruit_MQTT_Library/pull/28/files which lets me see all the changes at once...so forget what i said...learn something new every day!

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +0,0 @@
-language: c
```
Comment (toddtreece) [inline .travis.yml]: @stuthedew why did you remove the travis config?

Comment (stuthedew) [inline .travis.yml]: Oh, sorry about that. I deleted it because I initially planned on making this for just me and wasn't using it. It didn't occur to me to share it until after I removed it.

Comment (toddtreece) [inline .travis.yml]: no worries at all, and thanks for sharing your work! could you add the `.travis.yml` back to your fork when you get a chance? it would help us double check the pull req.

Comment (stuthedew) [inline .travis.yml]: Added back!

**Comment by stuthedew** [comment]
I got a Travis build to pass by adding appropriate platform skip files to the examples [Build Status](https://travis-ci.org/stuthedew/Adafruit_MQTT_Library).
The only possibly sketchy thing I did to get it to build is that I skipped the Arduino Zero in every example because it didn't build with any example. The errors were exactly the same as with your current master branch. Otherwise, everything seemed to build appropriately. I also added an arbitrary data publish example for the ethernet shield (in addition to ESP8266) that builds but is currently untested. I haven't merged the passing branch into the pull-request branch yet because I wasn't sure if that would be helpful or just confusing at this point. I can if you want me to. Just let me know.

P.S. If you hadn't guessed it already, this is my first real pull request. So sorry in advance for any pull-request faux pas!

**Comment by ladyada** [comment]
done!
